### PR TITLE
Factoring out parts of the serve protocol for Hydra to share

### PR DESCRIPTION
**Pull request by Ericson2314**
Hydra uses the lock argument differently than Nix, so we un-hard-code it.

See https://github.com/NixOS/hydra/pull/1165, which takes advantage of this (and #9560)

Depends on #9560

### CODE REVIEW

File: src/libstore/serve-protocol-impl.hh
Comment (thufschmitt) [inline src/libstore/serve-protocol-impl.hh:19]: Shouldn’t this be made a method of `BasicConnection` rather than a standalone function?

**Review by thufschmitt — COMMENTED** [review]
There might be a very good reason that I’m missing, but wouldn’t it be simpler to directly use the legacy ssh store in hydra? That would also make it possible to incrementally make hydra store-agnostic by extending the store API to fit its needs

File: src/libstore/serve-protocol-impl.hh
Comment (Ericson2314) [reply src/libstore/serve-protocol-impl.hh:19]: Sure. I put in a namespace like the `worker_proto` one we used earlier. Since this is sort of the "raw, secret" stuff, something about making it bare functions felt stylistically appropriate, but doesn't matter.

Comment (Ericson2314) [reply src/libstore/serve-protocol-impl.hh:19]: Eh, I sort of feel in an "OOP is not actually good" way that it is best to just make methods functions which are intrinsically about the data type itself, and this feels to me more like something one does with the data type.

Let me know what you want @thufschmitt. If you still disagree I'll change it, but I would like this to not drag on, so Hydra can keep on being converted.

**Comment by Ericson2314** [comment]
CI is not working because recent changes broke the macOS / clang builds.

**Comment by Ericson2314** [comment]
@thufschmitt This has been rebaced.

**Comment by Ericson2314** [comment]
Re what @thufschmitt wrote in https://github.com/NixOS/nix/pull/6134#issuecomment-1079199888:

Some steps like https://github.com/NixOS/hydra/pull/1180 are now done!

The big difference which makes it hard to reuse the `LegacySSHStore` as-is in Hydra is that every Nix operation grabs a new connection from the pool, whereas Hydra is resuing the same connection. And I am afraid this may make a difference for things like temp roots. So I think the sort of functions I am breaking out are still useful as a first more conserative step.

That said, I would be OK making `BasicClientConnection` part of `LegacySSHStore` and exposed `LegacySSHStore` as you say. (This would be similar to this PR but keeping more code in `legacy-ssh-store.cc`, where it is today, just moving it into these new functions.) How does that sound?

**Comment by Ericson2314** [comment]
team meeting notes:

- The "slow incremental migration" process is approved by the team (i.e. doing all these steps). That is on me torturing myself with many intermediate steps, shouldn't burden the team. When everything is done, and hydra isn't even using the legacy protocol any more, we should arrive at the same destination as if we did the "fast jump migration" process.

- Still blocked on @thufschmitt taking a look because his concerns from a year ago that may or may not apply.

**Comment by Ericson2314** [comment]
OK now this is rebased on top of the #9586 so we have the *option* of making these things part of `LegacySSHStore` for Hydra if @thufschmitt prefers!

**Comment by Ericson2314** [comment]
(in/after the meeting @thufschmitt said he didn't care about reviewing this one in particular.)

File: src/libstore/legacy-ssh-store.cc
Comment (roberth) [inline src/libstore/legacy-ssh-store.cc:71]: nitpick:

```suggestion
        // in.close(): Don't let the remote block on us not writing.
```

A bit more explicit, on 1 line.
We generally use `//` for this?

**Review by roberth — COMMENTED** [review]
It does look equivalent. Will approve when this is tested.

**Comment by Ericson2314** [comment]
OK added some tests!

File: src/libstore/serve-protocol-impl.hh
Comment (roberth) [inline src/libstore/serve-protocol-impl.hh:73]:
```suggestion
     * mechanisms.
```

File: src/libstore/serve-protocol-impl.hh
Comment (roberth) [inline src/libstore/serve-protocol-impl.hh:90]:
```suggestion
     * factored out serve protocol serializers with a
```

File: src/libstore/serve-protocol-impl.hh
Comment (roberth) [inline src/libstore/serve-protocol-impl.hh:106]:
```suggestion
     * factored out serve protocol serializers with a
```

File: src/libstore/serve-protocol-impl.hh
Comment (roberth) [inline src/libstore/serve-protocol-impl.hh:129]: If it doesn't go against a convention.

```suggestion
    void putBuildDerivationRequest(
```

File: src/libstore/serve-protocol-impl.hh
Comment (roberth) [inline src/libstore/serve-protocol-impl.hh:144]:
```suggestion
     * mechanisms.
```

File: tests/unit/libstore/serve-protocol.cc
Comment (roberth) [inline tests/unit/libstore/serve-protocol.cc:456]:
```suggestion
TEST_F(ServeProtoTest, handshake_client_truncated_replay_throws)
```

**Review by roberth — APPROVED** [review]

**Pull request merged**
